File: 169. Majority Element.py
from typing import List


class Solution:
    # def majorityElement(self, nums: List[int]) -> int:
    #     # time - O(n) space - O(n)
    #     counts = {}
    #
    #     for x in nums:
    #         temp = counts.get(x, 0) + 1
    #         if temp > len(nums) / 2:
    #             return x
    #         counts[x] = temp


    def majorityElement(self, nums: List[int]) -> int:
        # time - O(n) space - O(1)
        candidate, count = 0, 0

        for x in nums:
            if count == 0:
                candidate = x
                count = 1
            elif x == candidate:
                count += 1
            else:
                count -= 1
        return candidate


if "__main__" == __name__:
    solution = Solution()
    print(solution.majorityElement([3,3,2]))
    print(solution.majorityElement([2,2,2,1,1,1,2,]))

File: php/169. Majority Element.php
<?php

class Solution {

    /** time ON space O1
     * @param Integer[] $nums
     * @return Integer
     */
    function majorityElement($nums) {
        $freq = 0;
        $major_val = 0;

        foreach ($nums as $num) { 
            if ($freq == 0) {
                $major_val = $num;
                $freq = 1; 
            }  
            else if ($num == $major_val)
                $freq++;
            else $freq--;
        }
        return $major_val;
    }

    /** time - ON space ON 
     * @param Integer[] $nums
     * @return Integer
     */
    function majorityElement2($nums) {
        $freq = [];
        $major = $nums[0];

        foreach($nums as $num) {
            if (array_key_exists($num, $freq)){
                $freq[$num]++;
            }
            else {
                $freq[$num] = 1;
            }
            $major = $freq[$num] > $freq[$major] ? $num : $major;
        } 
        return $major;
    }
}

?>